lib/imagec/upload.go

```diff
@@ -0,0 +1,149 @@
+// See the License for the specific language governing permissi[ons and
+// limitations under the License.
+
+package imagec
+
+import (
+	"context"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/docker/docker/distribution/xfer"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// LayerUploader uploads layers
+type LayerUploader struct {
+	tm xfer.TransferManager
+}
+
+type uploadTransfer struct {
+	xfer.Transfer
+
+	err error
+}
+
+const (
+	maxUploadAttempts    = 5
+	maxConcurrentUploads = 3
+)
+
+// NewLayerUploader creates a new LayerUploader
+func NewLayerUploader() *LayerUploader {
+	return &LayerUploader{
+		tm: xfer.NewTransferManager(maxConcurrentUploads),
+	}
+}
+
+// UploadLayers starts the upload of all layers contained in the ImageC argument
+func (lum *LayerUploader) UploadLayers(ctx context.Context, ic *ImageC) error {
+	defer trace.End(trace.Begin(""))
+
+	var (
+		uploads        []*uploadTransfer
+		currTransfer   = make(map[string]*uploadTransfer)
+		sf             = streamformatter.NewJSONStreamFormatter()
+		progressOutput = &serialProgressOutput{
+			c:   make(chan prog, 100),
+			out: sf.NewProgressOutput(ic.Outstream, false),
+		}
+	)
+
+	go progressOutput.run()
+	defer progressOutput.stop()
+
+	pusher := ic.Pusher
+	log.Infof("There are %d layers to upload", len(pusher.streamMap))
+	for _, stream := range pusher.streamMap {
+		progress.Update(progressOutput, stream.layerID, "Preparing")
+
+		// Check if already uploading
+		if _, present := currTransfer[stream.layerID]; present {
+			continue
+		}
+
+		log.Infof("Making transfer function for %s...", stream.layerID)
+		xferFunc := lum.makeUploadFunc(ic, stream.layerID)
+		log.Infof("Starting transfer... function = %#v", xferFunc)
+		upload, watcher := lum.tm.Transfer(stream.layerID, xferFunc, progressOutput)
+		log.Infof("Tracking transfer...")
+		defer upload.Release(watcher)
+		uploads = append(uploads, upload.(*uploadTransfer))
+		currTransfer[stream.layerID] = upload.(*uploadTransfer)
+	}
+
+	for _, upload := range uploads {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-upload.Transfer.Done():
+			if upload.err != nil {
+				return upload.err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (lum *LayerUploader) makeUploadFunc(ic *ImageC, layerID string) xfer.DoFunc {
+	return func(progressChan chan<- progress.Progress, start <-chan struct{}, inactive chan<- struct{}) xfer.Transfer {
+		u := &uploadTransfer{
+			Transfer: xfer.NewTransfer(),
+		}
+
+		log.Infof("Getting stream reader from upload function")
+		aStream, reader, err := ic.Pusher.GetReaderForLayer(layerID)
+		if err != nil {
+			u.err = err
+			return u
+		}
+
+		go func() {
+			select {
+			case <-u.Transfer.Done():
+				reader.Close()
+			}
+		}()
+
+		go func() {
+			log.Info("About to upload...")
+			defer func() {
+				close(progressChan)
+				aStream.Close()
```

Comment:
aStream.Close will close reader, isn't this duplicate with line 119?
the reader is already set into aStream.layerfile, why do you still need a separate returned reader, and close those two separately.